presto-spi/src/main/java/com/facebook/presto/spi/relation/RowExpressionVisitor.java

```diff
@@ -27,8 +32,25 @@
 
     R visitSpecialForm(SpecialFormExpression specialForm, C context);
 
-    default R visitIntermediateFormExpression(IntermediateFormExpression intermediateFormExpression, C context)
+    // Default implementations for IntermediateFormRowExpression are provided below, as these only exist as
```

Comment:
I'm not sure if this is the right thing to do.  I think it would be better to have implementations handle IntermediateFormExpressions rather than worry about whether they are or are not in a situation that needs to handle it (or if they know they don't need to, then intentionally throw an exception).

Reply:
Agreed its good practice to let implementation handle this, so we get compile time errors than runtime errors.

Although I was hoping this to be an exception. There are 30+ implementations of RowExpressionVisitor. Given this form of expression lives for very small amount of time, and most(except 1-2) don't need to deal with them. I thought it best to not let them worry about this. Most of the rules should be added after the point we do this. What do you think?

Reply:
for the RowExpression visitors that exist, would it be a problem if we traversed into them (like DefaultRowExpressionTraversalVisitor). Maybe for the cases that don't want to inherit from that, we should do like PlanVisitor and AstVisitor, where everything default calls the visitor for it's superclass ending with visitPlan() for PlanVisitor or visitNode() for AstVisitor.  We can add a visitRowExpression() for RowExpressionVisitor and have that throw an exception by default?  That way we don't need to have special casing for the IntermediateFormExpression.  I do see that for expressions we have an implementation called DefaultExpressionTraversalVisitor that doesn't traverse into subqueries (again, don't understand why it's a problem), so we can add an implementation like that too for RowExpressions.

Reply:
Done. It's ready for another round of review!